trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/ACSLHandler.java

```diff
@@ -212,32 +222,80 @@ public Result visit(final IDispatcher main, final ACSLNode node) {
 	@Override
 	public Result visit(final IDispatcher main, final CodeAnnot node) {
 		if (node instanceof CodeAnnotStmt) {
-			final Check check;
-			if (mWitnessInvariantMode) {
-				check = new Check(Spec.WITNESS_INVARIANT);
-			} else {
-				check = new Check(Spec.ASSERT);
-			}
 			final ILocation loc = mLocationFactory.createACSLLocation(node);
-
-			ExpressionResult formula = (ExpressionResult) main
-					.dispatch(((Assertion) ((CodeAnnotStmt) node).getCodeStmt()).getFormula(), main.getAcslHook());
-
-			formula = mExprResultTransformer.transformSwitchRexIntToBool(formula, loc, main.getAcslHook());
-
 			final ExpressionResultBuilder resultBuilder = new ExpressionResultBuilder();
-			resultBuilder.addAllExceptLrValue(formula);
-
-			final AssertStatement assertStmt = new AssertStatement(loc, formula.getLrValue().getValue());
-			// TODO: Handle havoc statements
-			for (final Overapprox overapprItem : resultBuilder.getOverappr()) {
-				overapprItem.annotate(assertStmt);
+			final CodeStatement codeStmt = ((CodeAnnotStmt) node).getCodeStmt();
+			if (codeStmt instanceof Assertion) {
+				ExpressionResult formula =
+						(ExpressionResult) main.dispatch(((Assertion) codeStmt).getFormula(), main.getAcslHook());
+
+				formula = mExprResultTransformer.transformSwitchRexIntToBool(formula, loc, main.getAcslHook());
+
+				resultBuilder.addAllExceptLrValue(formula);
+
+				final AssertStatement assertStmt = new AssertStatement(loc, formula.getLrValue().getValue());
+				// TODO: Handle havoc statements
+				for (final Overapprox overapprItem : resultBuilder.getOverappr()) {
+					overapprItem.annotate(assertStmt);
+				}
+				resultBuilder.addStatement(assertStmt);
+				final List<HavocStatement> havocs = CTranslationUtil.createHavocsForAuxVars(formula.getAuxVars());
+				resultBuilder.addStatements(havocs);
+				final Check check;
+				if (mWitnessInvariantMode) {
+					check = new Check(Spec.WITNESS_INVARIANT);
+				} else {
+					check = new Check(Spec.ASSERT);
+				}
+				check.annotate(assertStmt);
+			}
+			if (codeStmt instanceof GhostUpdate) {
+				final GhostUpdate update = (GhostUpdate) codeStmt;
+				final ExpressionResult exprResult =
+						(ExpressionResult) main.dispatch(update.getExpr(), main.getAcslHook());
+				resultBuilder.addAllExceptLrValue(exprResult);
+				final SymbolTableValue stv = mSymboltable.findCSymbol(main.getAcslHook(), update.getIdentifier());
+				if (stv == null) {
+					throw new IncorrectSyntaxException(loc,
+							"Undeclared variable in ACSL expression: " + update.getIdentifier());
+				}
+				if (!stv.getBoogieName().startsWith(SFO.GHOST)) {
+					throw new IncorrectSyntaxException(loc,
+							"C variable " + update.getIdentifier() + " cannot be assigned in ghost statement.");

```

Comment:
Follow-up question: If yes, is this checked at the ghost declarations as well?

Reply:
For the declaration there should not be any shadowing. If we declare x as a program variable and a ghost variable, the first one is just translated to `x` (with some prefix), and the latter one to `#ghost~x`.
However, there could we shadowing if we have `x` as a program and ghost variable, and now `x` occurs in another ACSL statement (e.g. assert or ghost assignment on the right hand side), because wha `x` do we choose then? I just checked our implementation, we choose the program variable then. But I am not sure either, what the expected semantics are in ACSL for that case.

Reply:
We should also check the ACSL spec here. Intuitively, throwing an error at the declaration would be my preference -- but the precise handling might be hard in some cases, e.g. for global ACSL ghost variables and local C variables of the same name.

Reply:
I checked it (see 2.12 Ghost variables and statements), but I could not find anything related. I might however add a check for the declaration (check if the return of `FlatSymbolTable::findCSymbol` is null) and throw an exception for this case as well.